arrow-test/src/main/kotlin/arrow/test/laws/ConcurrentLaws.kt

```diff
@@ -199,9 +205,13 @@ object ConcurrentLaws {
       fx.concurrent {
         val mvar = MVar(a, this@releaseBracketIsNotCancelable).bind()
         val p = Promise.uncancelable<F, Unit>(this@releaseBracketIsNotCancelable).bind()
-        val task = p.complete(Unit)
-          .bracket(use = { never<Int>() }, release = { mvar.put(b) })
-        val (_, cancel) = task.fork(ctx).bind()
+        val task = just(Unit).bracket(
+          use = {
+            p.complete(Unit).flatMap { never<Int>() }
+          },
+          release = { mvar.put(b) }
+        )
+        val (_, cancel: Kind<F, Unit>) = task.fork().bind()
         p.get().bind()
```

Comment:
This change looks correct.

`MVar` is a single element blocking Queue, and we start with a filled `MVar` so every `put` will semantically block until `take`.

We use the `Promise` as a latch to make sure we `cancel` after `acquire` and during `use`.

`mvar.put(b)` in `release` will block since the `MVar` is already filled, and it will keep blocking until the last 2 statements empty the `MVar`.

This all happens after `cancelation` so we've correctly tested that `release` is not `cancelable`.

Reply:
The async jumping at the end seems unnecessary to me. Since everything is semantically blocking, and the system under test is being forked in a `Fiber`.

```kotlin
fun <F> Concurrent<F>.releaseBracketIsNotCancelable(EQ: Eq<Kind<F, Int>>, ctx: CoroutineContext) =
  forAll(Gen.int(), Gen.int()) { a, b ->
    fx.concurrent {
      val mvar = MVar(a, this@releaseBracketIsNotCancelable).bind() // Create filled MVar
      val p = Promise.uncancelable<F, Unit>(this@releaseBracketIsNotCancelable).bind() //Create use latch
      val task = just(Unit).bracket( // No-op acquire
        use = { p.complete(Unit).flatMap { never<Int>() } }, // Complete latch in `use` and never finish
        release = { mvar.put(b) } // Semantically block putting new value in MVar
      )
      val (_, cancel) = task.fork(ctx).bind() // / Fork the operation returning a `join, cancel` op
      p.get().bind() // Wait until `use`started
      cancel.fork(ctx).bind() // Cancel forked operation
      mvar.take().bind() // Take original value out of MVar
                                     // <-- Release blocks on `mvar.put` until here
      mvar.take().bind() // Take value put by release after `cancel`.
    }.equalUnderTheLaw(just(b), EQ)
  }
```

Reply:
Ok, Reaktive passes your variant as well. Will update.